sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineResponse, PipelineRequest, SendRequest } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { delay } from "../util/helpers";
+import { createClientLogger } from "@azure/logger";
+import { tryCreateSpan, tryProcessError, tryProcessResponse } from "./tracingPolicy";
+import { Span } from "@azure/core-tracing";
+import { getUserAgentValue } from "../util/userAgent";
+import { RetryStrategy, RetryStrategyState } from "../retryStrategies/retryStrategy";
+import { RestError } from "../restError";
+import { AbortError } from "@azure/abort-controller";
+
+const retryPolicyLogger = createClientLogger("core-rest-pipeline retryPolicy");
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * The programmatic identifier of the retryPolicy.
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(...strategies: RetryStrategy[]): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      let response: PipelineResponse | undefined;
+      let responseError: RestError | undefined;
+      let retryCount = -1;
+
+      const strategyState: RetryStrategyState[] = Array(strategies.length);
+      retryRequest: while (true) {
+        retryCount += 1;
+        response = undefined;
+        responseError = undefined;
+
+        let retrySpan: Span | undefined;
+        if (request.tracingOptions?.tracingContext) {
+          retrySpan = tryCreateSpan(request, getUserAgentValue(retryPolicyName));
+        }
+
+        try {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Attempting to send request`,
+            request.requestId
+          );
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+          tryProcessResponse(response, retrySpan);
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
```

Comment:
is it possible that `e` is not a `RestError`?

Reply:
good point! I will throw in that case